### DISCUSSION TITLE
The While clause of a Do While loop should exist in the loop's scope

### DISCUSSION THREAD

**Discussion by masonwheeler** [dotnet/csharplang#6340] [Language Ideas]
```
var data = GetData();
do {
   var foo = new Foo();
   Foo.ProcessSomeData(data);
}  while (!foo.Finished);
```

Formally, this does not work because `foo` is declared inside a new set of braces.  But intuitively it feels like it should, as the `while` clause is inherently concerned with what's going on inside the loop's body, particularly because in the absence of visibility into that scope, the only way to solve this problem is by scope pollution, leaking relevant data out into the outer scope:

```
var data = GetData();
var finished = false;
do {
   var foo = new Foo();
   Foo.ProcessSomeData(data);
   finished = foo.Finished;
}  while (!finished);
```

Are there any examples of existing code that would break by moving the `while` clause to be in scope for the loop body?  Intuitively it feels like the answer would be no, because the only way to make that a breaking change is if you had two different variables with the same name, one inside the loop body and the other outside, and the `while` clause referred to that variable.  But redeclaring a variable inside an inner scope that already exists in an outer scope is illegal in C#.  (It might be possible to do this with a field name, but fields and locals use different naming conventions that the IDE tries to dissuade people from violating, so this seems like a low-probability issue.)

**Comment by HaloFour** [comment]
Changing the scope would affect the lifetime which would impact how that variable is treated by closures as there'd be a single value, not one per iteration.  That would be a breaking change.

**Reply by masonwheeler** [reply]
> Changing the scope would affect the lifetime

I don't see how.  What exactly would it affect the lifetime of, and why?

**Reply by HaloFour** [reply]
The scope of the variable is within the block in which it is declared.  You'd have to expand the scope of the variable outside of the block in order to reference it from within the clauses of the `do`/`while` loop.  Variables scoped to those clauses would have a lifetime of the entire loop, not each iteration, so mutations would be observable by any closures capturing those variables.  This is how `for` loops work.

IMO it's slightly annoying but this is how all C-based languages have worked and the "workaround" is both simple and makes it clear what the lifecycle and scope of the variable is.

**Reply by masonwheeler** [reply]
I'm not asking to expand the scope of the variable; I'm asking to *move the while clause into the body's scope* as a point of convenience, even though it's not textually written inside the braces.

**Reply by HaloFour** [reply]
The body of the `do`/`while` loop is just an `embedded_statement`, if you use a `block` there that brings it's own scope, just as a `block` within that `block` would.  I'm not sure that moving the scope of the `while` clause so that it's a part of that scope makes sense, at least from a spec point of view.  But someone from the language team can speak a lot better about that than I could.

**Reply by CyrusNajmabadi** [reply]
It's certainly something we could do if we really felt it was warranted.  However, this would also be a breaking change that would likely hit real projects, so it practically cannot happen. 

Breaking changes have a very limited set if cases we allow them for, and a very limited window of time we'll accept for them. This issue very very likely would not meet the criteria necessary.

**Comment by PathogenDavid** [comment]
> (It might be possible to do this with a field name, but fields and locals use different naming conventions that the IDE tries to dissuade people from violating, so this seems like a low-probability issue.)

It's not uncommon for C# code to use camelCase for both fields and locals, but either way conventions are not a strong enough guarantee to change something fundamental like this.

The meaning of this code would be completely different with this proposed change:

```csharp
public class C
{
    public bool finished = false;
    
    public void M()
    {
        do
        {
            bool finished = true;
        } while (!finished);
    }
}
```

**Reply by masonwheeler** [reply]
Oh wow.  I just checked that, and the compiler doesn't even give a Message, let alone a Warning, that you're shadowing a field with that variable name.  It bugs you when you reuse an existing method name, so why not this? 😕

**Reply by CyrusNajmabadi** [reply]
Because this was allowed since 1.0 and became a normal and common idiom. 

We balance the value of warnings/errors against the value of the code that would be blocked. This is definitely common, understandable, and reasonable. So we don't warn for this.

**Reply by HaloFour** [reply]
Sounds like a good case for an analyzer, though.  IMO I always prefer to qualify fields with `this` if there happens to be a local variable of the same name in scope.

**Reply by CyrusNajmabadi** [reply]
I believe this is also one of the reasons why .net naming convention is to recommend that fields start with underscore. But that's just a convention of that group, and it's not something widely followed. Tons and tons of projects had and continue to have the convention of camelCase for fields.
